### ISSUE TITLE
When completing enum patterns, complete unqualifed enum names as well

### ISSUE THREAD

**Issue by matklad** [rust-analyzer/rust-analyzer#4014]
Given the following code:

```rust
enum Animal {
    Walrus,
    MantisShrimp
}

fn feed(animal: &mut Animal) {
    match animal {
        Wa /*cursor*/
    }
}
```

we should suggest `Animial::Walrus` completion. The relevant completion handler is [here](https://github.com/rust-analyzer/rust-analyzer/blob/179d9835351f3abab03634b50bcee94723cce148/crates/ra_ide/src/completion/complete_unqualified_path.rs). `ctx.sema.type_of_pat(..)` method can be used to get the type of the pattern.

**Comment by matklad** [comment]
See also https://internals.rust-lang.org/t/bring-enum-variants-in-scope-for-patterns/12104

**Comment by flodiebold** [comment]
We should do the same in expressions as well, actually! And to extend these type-based completions even further, we could e.g. suggest `X::new()` if the expected type is `X` (or any other static method on `X` returning `X`).

**Comment by djc** [comment]
In this context, would also be cool if the name resolution issue for importing the variants actually got fixed (see #3988).
